blog/2017-09-25-streaming-to-another-database.adoc

```diff
@@ -0,0 +1,222 @@
+
+.A General topology
+[#img-general]
+[caption="Figure 1: "]
+image::dbz-to-jdbc.svg[Scenario topology]
+
+To simplify the setup a little bit we will use only one Kafka Connect instance that will contain all connectors so the single instance will serve as an event producer and an event consumer.
+
+.A Simplified topology
+[#img-general]
+[caption="Figure 2: "]
+image::dbz-to-jdbc-simplified.svg[Scenario topology]
+
+== Configuration
+We will use this https://github.com/debezium/debezium-examples/tree/master/unwrap-smt[compose] for a fast deployment of the demo.
+The deployment consists of following images
+
+* Apache ZooKeeper
+* Apache Kafka
+* An enriched Kafka Connect image with changes
+** PostgreSQL JDBC driver placed into `/kafka/libs` directory
+** https://docs.confluent.io/current/connect/connect-jdbc/docs/index.html[Kafka Connect JDBC Connector] placed into `/kafka/connect/kafka-connect-jdbc` directory
+* Pre-populated MySQL used in our link:docs/tutorial[tutorial]
+* Empty PostgreSQL
+
+There are certain impedances between Debezium MySQL Connector and Kafka Connect JDBC connector
+
+* event format
+* delete records/tombstone records handling
+* topic naming
+
+We are going to go through all those issues in detail.
+
+=== Event format
+Debezium emits events in a complex format that contains two record values: a value of the row record before the change and value of the row record after the change is applied.
+Further metadata are also included and the result is so called `Envelope` format.
+
+[source,json,indent=0]
+----
+{
+	"op": "u",
+	"source": {
+		...
+	},
+	"ts_ms" : "...",
+	"before" : {
+		"field1" : "oldvalue1",
+		"field2" : "oldvalue2"
+	},
+	"after" : {
+		"field1" : "newvalue1",
+		"field2" : "newvalue2"
+	}
+}
+----
+
+Kafka Connect JDBC connector (as most of the others) emits outgoing and expects incoming events in a simple format that is generally a new row record.
+
+[source,json,indent=0]
+----
+{
+	"field1" : "newvalue1",
+	"field2" : "newvalue2"
+}
+----
+
+Debezium provides a bridge between those two formats in a form of a https://cwiki.apache.org/confluence/display/KAFKA/KIP-66%3A+Single+Message+Transforms+for+Kafka+Connect[single message transform].
+The `UnwrapFromEnvelope` transformation automatically extracts a new row record and thus effectively _flattens_ the complex record into a simple one consumable by other connectors.
+The SMT is applicable at both source and sink connectors but we generally recommend to apply it at sink connectors as then the records stored in Kafka will contain an associated context.
+In our example we apply the SMT at sink connector using configuration properties
+
+[source,indent=0]
+----
+        "transforms": "unwrap", 
+        "transforms.unwrap.type": "io.debezium.transforms.UnwrapFromEnvelope", 
+----
+
+=== Delete records
+When a record is deleted from source database Debezium creates two records: delete record and tombstone record (for compaction).
```

Comment:
> When the Debezium connector detects a row is deleted, it creates two event messages: a delete event and a tombstone message. The delete message has an envelope with the state of the deleted row in the `before` field, and an `after` field that is null. The tombstone message contains same key as the delete message, but the entire message value is null, and Kafka's log compaction utilizes this to know that it can remove any earlier messages with the same key.
> A number of sink connectors, including the Confluent's JDBC Sink Connector, are not expecting these messages and will instead fail if it sees either kind of message. The `UnwrapFromEnvelope` SMT will by default filter out both delete and tombstone records, though you can change this if you're using the SMT and want to keep one or both of these kinds of messages.

Reply:
Why is it btw. that the sink connector doesn't accept DELETEs, @rhauch? Seems like a useful addition?

Reply:
I think it helps to understand the context in which each connector was developed. Many of the sink connectors, including Confluent's Elasticsearch sink connector, were intended to capture the messages in one or more topics and write them out to Elasticsearch, regardless of whether those topics are compacted or not. And since deletes and tombstones only make sense on _compacted topics_, there was no special consideration for these kinds of messages.

I think the same is true for Confluent's JDBC sink connector, although that also was developed in conjunction with Confluent's JDBC _source connector_. My guess is that one more factor for not considering deletes and tombstones in the sink connector is that the source connector is not able to even detect deleted rows.

Although its certainly possible for event-driven applications to produce events that represent a removal, it is not necessary for these to be a tombstone. Really, CDC-like connectors may be one of the few source connectors that care to generate not only delete messages to capture the before state of the rows but also tombstone messages to signal Kafka that any previous messages with the same key could be compacted/removed.

Reply:
I'm not so much concerned about the technicality of tombstones or compaction, but rather that it seems not possible to convey the information that a record has been deleted to a sink such as Elasticsearch. Is the idea to only ever work with logical deletes, i.e. setting some "deleted marker" which would cause records to not be shown in results?